test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs

```diff
@@ -168,6 +368,97 @@ internal HttpMessage CreateCollapseGetRequest(RequestConditions requestCondition
             return message;
         }
 
+        internal HttpMessage CreateMissIfNonMatchGetRequest(RequestConditions requestConditions, RequestContext context)
+        {
+            var message = _pipeline.CreateMessage(context);
+            var request = message.Request;
+            request.Method = RequestMethod.Get;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/RequestConditionCollapse/missIfNonMatch", false);
+            request.Uri = uri;
+            if (requestConditions != null)
+            {
+                if (requestConditions.IfNoneMatch != null) throw new NotSupportedException($"IfNoneMatch head is not supported here");
+                request.Headers.Add(requestConditions, "R");
+            }
+            message.ResponseClassifier = ResponseClassifier200.Instance;
+            return message;
+        }
+
+        internal HttpMessage CreateMissIfMatchGetRequest(RequestConditions requestConditions, RequestContext context)
+        {
+            var message = _pipeline.CreateMessage(context);
+            var request = message.Request;
+            request.Method = RequestMethod.Get;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/RequestConditionCollapse/missIfMatch", false);
+            request.Uri = uri;
+            if (requestConditions != null)
+            {
+                if (requestConditions.IfMatch != null) throw new NotSupportedException($"IfMatch head is not supported here");
```

Comment:
Supported values are the feature of parameter, not header. Hence this check should happen in `MissIfMatchGet`/`MissIfMatchGetAsync` methods, and exception type will be `ArgumentException`. Ideally, I'd like to have a method in `Azure.Core.Argument`, similar to `Argument.AssertNotNullOrEmpty`, something like:
```cs
public static void AssertHasOnlySupportedHeaders(RequestConditions? requestConditions, string paramName, params string[] supportedHeaders)
{
    if (requestConditions != null)
    {
        AssertHasNoUnsupportedHeader(requestConditions.IfMatch, paramName, supportedHeaders, HttpHeader.Names.IfMatch);
        AssertHasNoUnsupportedHeader(requestConditions.IfNoneMatch, paramName, supportedHeaders, HttpHeader.Names.IfNoneMatch);
        AssertHasNoUnsupportedHeader(requestConditions.IfModifiedSince, paramName, supportedHeaders, HttpHeader.Names.IfModifiedSince);
        AssertHasNoUnsupportedHeader(requestConditions.IfUnmodifiedSince, paramName, supportedHeaders, HttpHeader.Names.IfUnmodifiedSince);
    }
}

private static void AssertHasNoUnsupportedHeader<T>(T? value, string paramName, string[] supportedHeaders, string unsupportedHeader) where T : struct
{
    if (value.HasValue && !supportedHeaders.Contains(unsupportedHeader))
    {
        throw new ArgumentException($"{unsupportedHeader} header is not supported. Supported headers are: {string.Join(", ", supportedHeaders)}.", paramName);
    }
}
```
```cs
public virtual async Task<Response> MissIfMatchGetAsync(RequestConditions requestConditions = null, RequestContext context = null) 
{
    Argument.AssertHasOnlySupportedHeaders(requestConditions, nameof(requestConditions), "If-None-Match", "If-Modified-Since");
```

This approach would have several benefits:
- testability
- less generated code
- no need to regenerate old code if another property is added to `RequestConditions`

Reply:
The question however is how to add validation information to the `AutoRest.CSharp.Output.Models.Shared.Parameter`, since we don't have any support for custom validation. Let me think about it.

Reply:
Yes, currently we have not put parameter validation in the rest Method, and currently we have put some validation in Create****Request, such as check if a header is NULL when try to add the Head to HTTP Request. We can put the validation either location, but we need to consistent.

Reply:
I'd like to simplify what goes into Argument.cs so that the generated code looks more like this:

```
public virtual Response DoThing(RequestConditions conditions, RequestContext context = default)
{
  Argument.AssertNull(requestConditions.IfMatch, nameof(requestConditions, "Service does not support the If-Match header for this operation.");
  // the rest of the service call
}

Reply:
hi @annelo-msft 
When there are more than one request condition need to check, the Argument.AssertNull will be call more than once.
e..g 

public virtual Response DoThing(RequestConditions conditions, RequestContext context = default)
{
    Argument.AssertNull(requestConditions.IfMatch, nameof(requestConditions), "Service does not support the If-Match header for this operation.");
  Argument.AssertNull(requestConditions.IfNoneMatch, nameof(requestConditions), "Service does not support the If-None-Match header for this operation.");
    // the rest of the service call
}

The worst case, the Argument.AssertNull will call three times.
Is it acceptable?

Reply:
I think this is a very edge case where we have these asserts in the first place.  I think simplicity of Azure.Core APIs and readability of the code should be the driving factor, and we can revisit the perf implications if we find we have a problem in the future.  This would give us a strong motivating reason to write more complex APIs into the Assert class.

We will not add properties to RequestConditions often -- this type represents [HTTP conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests), and has [four of the five headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#conditional_headers) represented.  We haven't yet found a case in Azure that needed the fifth - we would only add it if we did, and then there would be no further headers to add.

And @AlexanderSher, I don't understand the argument about having to regenerate code -- don't we always regenerate code when we make changes to the generator?

Reply:
When we add a new field into RequestConditions type, then we need to also check if the new field is set or not. In that way, we need to regenerated the existing SDK client libraries to reflect it. 
But if we use AssertHasOnlySupportedHeaders, we just need to update the method in Azure.Core, the existing SDK client libraries does not updated.

Reply:
Yes, RequestConditions type is for HTTP conditional requests, so it is rare or maybe not to be updated (such as add or remove an field), so we may not need to consider this scenario current. Right? @annelo-msft

Reply:
We'll never remove a field, because that would be a breaking change, and there is only one field left to add, so we would only ever regenerate once.  So, yes, I wouldn't worry about the possible addition - we regenerate the library code fairly often anyway.

Reply:
I will implement it this way.

Reply:
Thanks, @chunyu3!